app/clients/TrackDownloadsClient.ts

```diff
@@ -0,0 +1,254 @@
+          state.progress = totalBytesWritten / totalBytesExpectedToWrite
+        }
+        this.notify(track.trackId)
+      }
+
+      const downloadResumable = FileSystem.createDownloadResumable(
+        track.webUri,
+        getTemporaryFileUri(track),
+        // NOTE: Make sure to account for FileSystemDownloadResult.md5 not
+        // existing if you change this to `false` or remove it.
+        { md5: true },
+        callback,
+      )
+      downloadResumable.downloadAsync().then(
+        (r) => {
+          this.handleDownloaded(track, r)
+        },
+        (e) => {
+          this.handleDownloadFailed(track.trackId, e)
+        },
+      )
+      this.cachedDownloadState.set(track.trackId, {
+        status: "DOWNLOADING",
+        progress: 0,
+        downloadResumable,
+      })
+      this.notify(track.trackId)
+    } else if (currentState.status === "DOWNLOAD_PAUSED") {
+      currentState.downloadResumable.resumeAsync().then(
+        (r) => {
+          this.handleDownloaded(track, r)
+        },
+        (e) => {
+          this.handleDownloadFailed(track.trackId, e)
+        },
+      )
+
+      this.cachedDownloadState.set(track.trackId, {
+        status: "DOWNLOADING",
+        progress: currentState.progress,
+        downloadResumable: currentState.downloadResumable,
+      })
+      this.notify(track.trackId)
+    }
+  }
+
+  pauseDownload(trackId: string): void {
+    const currentState = this.fetchDownloadState(trackId)
+    if (currentState.status === "DOWNLOADING") {
+      currentState.downloadResumable.pauseAsync().then(
+        () => {
+          this.cachedDownloadState.set(trackId, {
+            status: "DOWNLOAD_PAUSED",
+            progress: currentState.progress,
+            downloadResumable: currentState.downloadResumable,
+          })
+          this.notify(trackId)
+        },
+        // TODO: Handle this case.
+        (e) => console.log("pause failed", e),
+      )
+    }
+  }
+
+  clearDownload(track: Track): void {
+    FileSystem.deleteAsync(getTrackFileUri(track)).then(
+      () => {
+        this.cachedDownloadState.set(track.trackId, {
+          status: "NOT_DOWNLOADED",
+        })
+        this.notify(track.trackId)
+      },
+      (error) => {
+        console.error(`Failed to delete temporary file of track ${track.trackId}`, error)
+      },
+    )
+  }
+
+  subscribe(trackId: string, callback: () => void): () => void {
+    const entry = this.listeners.get(trackId)
+
+    if (entry) {
+      entry.push(callback)
+    } else {
+      this.listeners.set(trackId, [callback])
+    }
+
+    const unsubscribe = () => {
+      const entries = this.listeners.get(trackId) ?? []
+
+      this.listeners.set(
+        trackId,
+        entries.filter((e) => e !== callback),
+      )
+    }
+
+    return unsubscribe
+  }
+
+  protected notify(trackId: string) {
+    const entries = this.listeners.get(trackId) ?? []
+
+    entries.forEach((callback) => callback())
```

Comment:
I'm not sure. I'm currently leaning on the side of "let it fail, it's a programmer error anyway, nothing we can recover from" and adding [error boundaries](https://reactjs.org/docs/error-boundaries.html)/[global exception handlers](https://github.com/a7ul/react-native-exception-handler) to prevent the entire app from crashing and/or to send us error reports.